File: packages/ui-swiftui/Sources/KinetixUI/Theme.swift
//
// Theme.swift — the KinetixTheme wrapper.
//
// The SwiftUI analogue of packages/ui-compose's KinetixColorScheme /
// KinetixTheme: a small semantic colour set every `Kinetix*` view reads
// from `@Environment(\.kinetixColors)`, plus a `KinetixTheme { … }`
// wrapper that picks light or dark off the system `colorScheme`.
//
// The colour VALUES come from the generated, vendored
// KinetixColorsSwiftUI / KinetixColorsSwiftUIDark enums
// (KinetixColorsSwiftUI.swift / .dark.swift — do not hand-edit those;
// re-run `node scripts/vendor-swiftui-tokens.mjs` after `pnpm build:tokens`).
// Those are SwiftUI `Color`, not `UIColor`: UIColor is UIKit-only and won't
// compile on macOS, so the token engine emits this parallel Color set for
// the library (the existing KinetixColors.swift / Theme.swift UIColor
// output is untouched, for existing consumers).
//

import SwiftUI

/// The semantic colour set a KinetixUI component resolves to — the same
/// handful of tokens every variant in the React source
/// (packages/ui/src/components) maps to (bg-primary, text-foreground, …).
/// Matches `KinetixColors` in packages/ui-compose/.../Theme.kt field-for-field.
public struct KinetixColors {
    public let primary: Color
    public let primaryForeground: Color
    public let secondary: Color
    public let secondaryForeground: Color
    public let destructive: Color
    public let destructiveForeground: Color
    public let foreground: Color
    public let background: Color
    public let border: Color
    public let input: Color
    public let ring: Color
    public let muted: Color
    public let mutedForeground: Color
    public let accent: Color
    public let accentForeground: Color
    public let tertiary: Color
    public let tertiaryForeground: Color
    public let warning: Color
    public let warningForeground: Color
    public let card: Color
    public let cardForeground: Color
    public let success: Color
    public let successForeground: Color
    public let info: Color
    public let infoForeground: Color
    public let popover: Color
    public let popoverForeground: Color
    /// The 5-stop categorical chart palette (`--chart-1` … `--chart-5`).
    public let chart: [Color]

    public init(
        primary: Color,
        primaryForeground: Color,
        secondary: Color,
        secondaryForeground: Color,
        destructive: Color,
        destructiveForeground: Color,
        foreground: Color,
        background: Color,
        border: Color,
        input: Color,
        ring: Color,
        muted: Color,
        mutedForeground: Color,
        accent: Color,
        accentForeground: Color,
        tertiary: Color,
        tertiaryForeground: Color,
        warning: Color,
        warningForeground: Color,
        card: Color,
        cardForeground: Color,
        success: Color,
        successForeground: Color,
        info: Color,
        infoForeground: Color,
        popover: Color,
        popoverForeground: Color,
        chart: [Color]
    ) {
        self.primary = primary
        self.primaryForeground = primaryForeground
        self.secondary = secondary
        self.secondaryForeground = secondaryForeground
        self.destructive = destructive
        self.destructiveForeground = destructiveForeground
        self.foreground = foreground
        self.background = background
        self.border = border
        self.input = input
        self.ring = ring
        self.muted = muted
        self.mutedForeground = mutedForeground
        self.accent = accent
        self.accentForeground = accentForeground
        self.tertiary = tertiary
        self.tertiaryForeground = tertiaryForeground
        self.warning = warning
        self.warningForeground = warningForeground
        self.card = card
        self.cardForeground = cardForeground
        self.success = success
        self.successForeground = successForeground
        self.info = info
        self.infoForeground = infoForeground
        self.popover = popover
        self.popoverForeground = popoverForeground
        self.chart = chart
    }
}

public extension KinetixColors {
    /// Built from the generated light enum — `pnpm build:tokens`.
    static let light = KinetixColors(
        primary: KinetixColorsSwiftUI.primary,
        primaryForeground: KinetixColorsSwiftUI.primaryForeground,
        secondary: KinetixColorsSwiftUI.secondary,
        secondaryForeground: KinetixColorsSwiftUI.secondaryForeground,
        destructive: KinetixColorsSwiftUI.destructive,
        destructiveForeground: KinetixColorsSwiftUI.destructiveForeground,
        foreground: KinetixColorsSwiftUI.foreground,
        background: KinetixColorsSwiftUI.background,
        border: KinetixColorsSwiftUI.border,
        input: KinetixColorsSwiftUI.input,
        ring: KinetixColorsSwiftUI.ring,
        muted: KinetixColorsSwiftUI.muted,
        mutedForeground: KinetixColorsSwiftUI.mutedForeground,
        accent: KinetixColorsSwiftUI.accent,
        accentForeground: KinetixColorsSwiftUI.accentForeground,
        tertiary: KinetixColorsSwiftUI.tertiary,
        tertiaryForeground: KinetixColorsSwiftUI.tertiaryForeground,
        warning: KinetixColorsSwiftUI.warning,
        warningForeground: KinetixColorsSwiftUI.warningForeground,
        card: KinetixColorsSwiftUI.card,
        cardForeground: KinetixColorsSwiftUI.cardForeground,
        success: KinetixColorsSwiftUI.success,
        successForeground: KinetixColorsSwiftUI.successForeground,
        info: KinetixColorsSwiftUI.info,
        infoForeground: KinetixColorsSwiftUI.infoForeground,
        popover: KinetixColorsSwiftUI.popover,
        popoverForeground: KinetixColorsSwiftUI.popoverForeground,
        chart: [
            KinetixColorsSwiftUI.chart1,
            KinetixColorsSwiftUI.chart2,
            KinetixColorsSwiftUI.chart3,
            KinetixColorsSwiftUI.chart4,
            KinetixColorsSwiftUI.chart5,
        ]
    )

    /// Built from the generated dark enum — the real dark pass, same as the
    /// web `.dark` selector (nothing else native consumes a dark pass yet).
    static let dark = KinetixColors(
        primary: KinetixColorsSwiftUIDark.primary,
        primaryForeground: KinetixColorsSwiftUIDark.primaryForeground,
        secondary: KinetixColorsSwiftUIDark.secondary,
        secondaryForeground: KinetixColorsSwiftUIDark.secondaryForeground,
        destructive: KinetixColorsSwiftUIDark.destructive,
        destructiveForeground: KinetixColorsSwiftUIDark.destructiveForeground,
        foreground: KinetixColorsSwiftUIDark.foreground,
        background: KinetixColorsSwiftUIDark.background,
        border: KinetixColorsSwiftUIDark.border,
        input: KinetixColorsSwiftUIDark.input,
        ring: KinetixColorsSwiftUIDark.ring,
        muted: KinetixColorsSwiftUIDark.muted,
        mutedForeground: KinetixColorsSwiftUIDark.mutedForeground,
        accent: KinetixColorsSwiftUIDark.accent,
        accentForeground: KinetixColorsSwiftUIDark.accentForeground,
        tertiary: KinetixColorsSwiftUIDark.tertiary,
        tertiaryForeground: KinetixColorsSwiftUIDark.tertiaryForeground,
        warning: KinetixColorsSwiftUIDark.warning,
        warningForeground: KinetixColorsSwiftUIDark.warningForeground,
        card: KinetixColorsSwiftUIDark.card,
        cardForeground: KinetixColorsSwiftUIDark.cardForeground,
        success: KinetixColorsSwiftUIDark.success,
        successForeground: KinetixColorsSwiftUIDark.successForeground,
        info: KinetixColorsSwiftUIDark.info,
        infoForeground: KinetixColorsSwiftUIDark.infoForeground,
        popover: KinetixColorsSwiftUIDark.popover,
        popoverForeground: KinetixColorsSwiftUIDark.popoverForeground,
        chart: [
            KinetixColorsSwiftUIDark.chart1,
            KinetixColorsSwiftUIDark.chart2,
            KinetixColorsSwiftUIDark.chart3,
            KinetixColorsSwiftUIDark.chart4,
            KinetixColorsSwiftUIDark.chart5,
        ]
    )
}

private struct KinetixColorsKey: EnvironmentKey {
    static let defaultValue = KinetixColors.light
}

public extension EnvironmentValues {
    var kinetixColors: KinetixColors {
        get { self[KinetixColorsKey.self] }
        set { self[KinetixColorsKey.self] = newValue }
    }
}

/// Wrap a screen (or a preview) in `KinetixTheme { … }` so every `Kinetix*`
/// view resolves to the right light/dark value set. Defaults to the system
/// `colorScheme`, the same behaviour as the web `next-themes` "system" mode
/// and Compose's `KinetixTheme(darkTheme: isSystemInDarkTheme())`.
///
/// ```swift
/// KinetixTheme {
///     KinetixButton(action: {}) { Text("Save") }
/// }
/// ```
public struct KinetixTheme<Content: View>: View {
    @Environment(\.colorScheme) private var colorScheme
    private let content: Content

    public init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }

    public var body: some View {
        content.environment(\.kinetixColors, colorScheme == .dark ? .dark : .light)
    }
}

File: packages/ui-flutter/lib/src/theme.dart
import 'package:flutter/widgets.dart';

import 'kinetix_color_scheme.dart';
import 'kinetix_color_scheme.dark.dart';

/// The semantic colour set every `Kinetix*` widget resolves to — the same
/// handful of tokens each variant in `packages/ui/src/components` maps to
/// (bg-primary, text-foreground, …). Matches `KinetixColors` in the
/// Compose / SwiftUI ports field-for-field.
///
/// The values come from the generated, vendored [KinetixColorScheme] /
/// [KinetixColorSchemeDark] classes (`kinetix_color_scheme.dart` /
/// `.dark.dart` — do not hand-edit; re-run
/// `node scripts/vendor-flutter-tokens.mjs` after `pnpm build:tokens`).
@immutable
class KinetixColors {
  const KinetixColors({
    required this.primary,
    required this.primaryForeground,
    required this.secondary,
    required this.secondaryForeground,
    required this.destructive,
    required this.destructiveForeground,
    required this.foreground,
    required this.background,
    required this.border,
    required this.input,
    required this.ring,
    required this.muted,
    required this.mutedForeground,
    required this.accent,
    required this.accentForeground,
    required this.tertiary,
    required this.tertiaryForeground,
    required this.warning,
    required this.warningForeground,
    required this.card,
    required this.cardForeground,
    required this.success,
    required this.successForeground,
    required this.info,
    required this.infoForeground,
    required this.popover,
    required this.popoverForeground,
    required this.chart,
  });

  final Color primary;
  final Color primaryForeground;
  final Color secondary;
  final Color secondaryForeground;
  final Color destructive;
  final Color destructiveForeground;
  final Color foreground;
  final Color background;
  final Color border;
  final Color input;
  final Color ring;
  final Color muted;
  final Color mutedForeground;
  final Color accent;
  final Color accentForeground;
  final Color tertiary;
  final Color tertiaryForeground;
  final Color warning;
  final Color warningForeground;
  final Color card;
  final Color cardForeground;
  final Color success;
  final Color successForeground;
  final Color info;
  final Color infoForeground;
  final Color popover;
  final Color popoverForeground;

  /// The 5-stop categorical chart palette (`--chart-1` … `--chart-5`).
  final List<Color> chart;

  /// Built from the generated light class — `pnpm build:tokens`.
  static const KinetixColors light = KinetixColors(
    primary: KinetixColorScheme.primary,
    primaryForeground: KinetixColorScheme.primaryForeground,
    secondary: KinetixColorScheme.secondary,
    secondaryForeground: KinetixColorScheme.secondaryForeground,
    destructive: KinetixColorScheme.destructive,
    destructiveForeground: KinetixColorScheme.destructiveForeground,
    foreground: KinetixColorScheme.foreground,
    background: KinetixColorScheme.background,
    border: KinetixColorScheme.border,
    input: KinetixColorScheme.input,
    ring: KinetixColorScheme.ring,
    muted: KinetixColorScheme.muted,
    mutedForeground: KinetixColorScheme.mutedForeground,
    accent: KinetixColorScheme.accent,
    accentForeground: KinetixColorScheme.accentForeground,
    tertiary: KinetixColorScheme.tertiary,
    tertiaryForeground: KinetixColorScheme.tertiaryForeground,
    warning: KinetixColorScheme.warning,
    warningForeground: KinetixColorScheme.warningForeground,
    card: KinetixColorScheme.card,
    cardForeground: KinetixColorScheme.cardForeground,
    success: KinetixColorScheme.success,
    successForeground: KinetixColorScheme.successForeground,
    info: KinetixColorScheme.info,
    infoForeground: KinetixColorScheme.infoForeground,
    popover: KinetixColorScheme.popover,
    popoverForeground: KinetixColorScheme.popoverForeground,
    chart: <Color>[
      KinetixColorScheme.chart1,
      KinetixColorScheme.chart2,
      KinetixColorScheme.chart3,
      KinetixColorScheme.chart4,
      KinetixColorScheme.chart5,
    ],
  );

  /// Built from the generated dark class — the real dark pass, same as the
  /// web `.dark` selector.
  static const KinetixColors dark = KinetixColors(
    primary: KinetixColorSchemeDark.primary,
    primaryForeground: KinetixColorSchemeDark.primaryForeground,
    secondary: KinetixColorSchemeDark.secondary,
    secondaryForeground: KinetixColorSchemeDark.secondaryForeground,
    destructive: KinetixColorSchemeDark.destructive,
    destructiveForeground: KinetixColorSchemeDark.destructiveForeground,
    foreground: KinetixColorSchemeDark.foreground,
    background: KinetixColorSchemeDark.background,
    border: KinetixColorSchemeDark.border,
    input: KinetixColorSchemeDark.input,
    ring: KinetixColorSchemeDark.ring,
    muted: KinetixColorSchemeDark.muted,
    mutedForeground: KinetixColorSchemeDark.mutedForeground,
    accent: KinetixColorSchemeDark.accent,
    accentForeground: KinetixColorSchemeDark.accentForeground,
    tertiary: KinetixColorSchemeDark.tertiary,
    tertiaryForeground: KinetixColorSchemeDark.tertiaryForeground,
    warning: KinetixColorSchemeDark.warning,
    warningForeground: KinetixColorSchemeDark.warningForeground,
    card: KinetixColorSchemeDark.card,
    cardForeground: KinetixColorSchemeDark.cardForeground,
    success: KinetixColorSchemeDark.success,
    successForeground: KinetixColorSchemeDark.successForeground,
    info: KinetixColorSchemeDark.info,
    infoForeground: KinetixColorSchemeDark.infoForeground,
    popover: KinetixColorSchemeDark.popover,
    popoverForeground: KinetixColorSchemeDark.popoverForeground,
    chart: <Color>[
      KinetixColorSchemeDark.chart1,
      KinetixColorSchemeDark.chart2,
      KinetixColorSchemeDark.chart3,
      KinetixColorSchemeDark.chart4,
      KinetixColorSchemeDark.chart5,
    ],
  );
}

/// Inherited holder for the active [KinetixColors]. Wrap a screen (or the
/// whole app) in `KinetixTheme(child: …)`; every `Kinetix*` widget below
/// reads it via `KinetixTheme.of(context)`. Defaults to the platform
/// brightness — pass [brightness] to force light or dark, the same as the
/// Compose port's `KinetixTheme(darkTheme: …)`.
class KinetixTheme extends InheritedWidget {
  const KinetixTheme({
    super.key,
    this.brightness,
    required super.child,
  });

  final Brightness? brightness;

  Brightness _brightnessFor(BuildContext context) =>
      brightness ??
      MediaQuery.maybeOf(context)?.platformBrightness ??
      WidgetsBinding.instance.platformDispatcher.platformBrightness;

  KinetixColors _colorsFor(BuildContext context) =>
      _brightnessFor(context) == Brightness.dark ? KinetixColors.dark : KinetixColors.light;

  /// The active semantic colour set for [context].
  static KinetixColors of(BuildContext context) {
    final theme = context.dependOnInheritedWidgetOfExactType<KinetixTheme>();
    assert(
      theme != null,
      'No KinetixTheme found in context. Wrap your app in KinetixTheme(child: …).',
    );
    return theme!._colorsFor(context);
  }

  @override
  bool updateShouldNotify(KinetixTheme oldWidget) => oldWidget.brightness != brightness;
}
